### PR TITLE
Fixed #615: There was a way Glide could get stuck in a resolver loop

### LINKED ISSUE #615: Glide seems to hang with high CPU importing kubernetes API
Create:

$GOPATH/src/example.com/example/main.go:
```
package main

import (
	"k8s.io/kubernetes/pkg/client/restclient"
	client "k8s.io/kubernetes/pkg/client/unversioned"
)

func main() {
	var config *restclient.Config
	client.New(config)
}
```

Now run:
```
cd $GOPATH/src/example.com/example
glide create --non-interactive
glide install
```

It will make slow progress for a while but eventually seems to hits a wall and seems to make no progress.  It's using about 1 full CPU core while resident size increases only very slowly.

I'm using go 1.6 on Ubuntu Xenial and glide 0.12.1 from ppa.

Output looks like:
```
$ glide init --non-interactive
[INFO]	Generating a YAML configuration file and guessing the dependencies
[INFO]	Attempting to import from other package managers (use --skip-import to skip)
[INFO]	Scanning code to look for dependencies
[INFO]	--> Found reference to k8s.io/kubernetes/pkg/client/restclient
[INFO]	--> Adding sub-package pkg/client/unversioned to k8s.io/kubernetes
[INFO]	Writing configuration file (glide.yaml)
[INFO]	You can now edit the glide.yaml file. Consider:
[INFO]	--> Using versions and ranges. See https://glide.sh/docs/versions/
[INFO]	--> Adding additional metadata. See https://glide.sh/docs/glide.yaml/
[INFO]	--> Running the config-wizard command to improve the versions in your configuration

$ glide install
[INFO]	Lock file (glide.lock) does not exist. Performing update.
[INFO]	Downloading dependencies. Please wait...
[INFO]	--> Fetching updates for k8s.io/kubernetes.
[INFO]	Resolving imports
[INFO]	--> Fetching updates for github.com/golang/glog.
[INFO]	--> Fetching updates for github.com/google/cadvisor.
[INFO]	--> Fetching updates for github.com/davecgh/go-spew.
[INFO]	--> Fetching updates for github.com/ugorji/go.
[INFO]	--> Fetching updates for golang.org/x/net.
[INFO]	--> Fetching updates for github.com/go-openapi/spec.
[INFO]	--> Fetching updates for github.com/gogo/protobuf.
[INFO]	--> Fetching updates for github.com/google/gofuzz.
[INFO]	--> Fetching updates for github.com/juju/ratelimit.
[INFO]	--> Fetching updates for github.com/blang/semver.
[INFO]	--> Fetching updates for github.com/emicklei/go-restful.
[INFO]	--> Fetching updates for github.com/spf13/pflag.
[INFO]	--> Fetching updates for gopkg.in/inf.v0.
[INFO]	--> Fetching updates for github.com/pborman/uuid.
[INFO]	--> Fetching updates for github.com/go-openapi/jsonpointer.
[INFO]	--> Fetching updates for github.com/go-openapi/jsonreference.
[INFO]	--> Fetching updates for github.com/go-openapi/swag.
[INFO]	--> Fetching updates for github.com/docker/distribution.
[INFO]	--> Fetching updates for golang.org/x/oauth2.
[INFO]	--> Fetching updates for k8s.io/client-go.
[INFO]	--> Fetching updates for github.com/coreos/go-oidc.
[INFO]	--> Fetching updates for github.com/ghodss/yaml.
[INFO]	--> Fetching updates for github.com/PuerkitoBio/purell.
[INFO]	--> Fetching updates for github.com/mailru/easyjson.
[INFO]	--> Fetching updates for google.golang.org/appengine.
[INFO]	--> Fetching updates for cloud.google.com/go.
[INFO]	--> Fetching updates for github.com/coreos/pkg.
[INFO]	--> Fetching updates for github.com/jonboulle/clockwork.
[INFO]	--> Fetching updates for github.com/cloudfoundry-incubator/candiedyaml.
[INFO]	--> Fetching updates for github.com/PuerkitoBio/urlesc.
[INFO]	--> Fetching updates for golang.org/x/text.
```

### PR DESCRIPTION
**Pull request by mattfarina**
The resolver loop caused Kubernetes imports to fail.

### CODE REVIEW

**Comment by seh** [comment]
Not likely related, but I ran into this problem running _glide install_ using this patch:
```text
[INFO]	--> Fetching cloud.google.com/go/cmd/go-cloud-debug-agent/internal/breakpoints.
[WARN]	Unable to checkout cloud.google.com/go/cmd/go-cloud-debug-agent/internal/breakpoints
[ERROR]	Error looking for cloud.google.com/go/cmd/go-cloud-debug-agent/internal/breakpoints: https://cloud.google.com/go/cmd/go-cloud-debug-agent/internal/breakpoints Not Found
```
Later, I ran into this problem, which I didn't see with Glide version 0.11.1:
```text
[ERROR]	Error scanning github.com/go-kit/kit/service: open /Users/seh/.glide/cache/src/https-github.com-go-kit-kit/service: no such file or directory
[ERROR]	This error means the referenced package was not found.
[ERROR]	Missing file or directory errors usually occur when multiple packages
[ERROR]	share a common dependency and the first reference encountered by the scanner
[ERROR]	sets the version to one that does not contain a subpackage needed required
[ERROR]	by another package that uses the shared dependency. Try setting a
[ERROR]	version in your glide.yaml that works for all packages that share this
[ERROR]	dependency.
[ERROR]	Error scanning google.golang.org/api/internal: open /Users/seh/.glide/cache/src/https-google.golang.org-api/internal: no such file or directory
[ERROR]	This error means the referenced package was not found.
[ERROR]	Missing file or directory errors usually occur when multiple packages
[ERROR]	share a common dependency and the first reference encountered by the scanner
[ERROR]	sets the version to one that does not contain a subpackage needed required
[ERROR]	by another package that uses the shared dependency. Try setting a
[ERROR]	version in your glide.yaml that works for all packages that share this
[ERROR]	dependency.
```
My code doesn't make reference to either of those packages.

**Comment by seh** [comment]
My second attempt yielded about ten more cases of this error:
> This error means the referenced package was not found.

**Comment by timoreimann** [comment]
@seh I _think_ I have seen this error before and fixed it by clearing my cache: `glide cache-clear`.

**Comment by technosophos** [comment]
@mattfarina Yup, that is probably it.

**Comment by seh** [comment]
How does one know when it's time to run _glide [cache-clear](https://github.com/Masterminds/glide/blob/master/action/cache.go#L11)_?
I assumed that _glide install_ would trigger _glide update_ (for lack of a _glide.lock_ file), which would update the cache. Is that not the case?

File: dependency/resolver.go [hunk 1]
```diff
@@ -460,12 +460,19 @@ func (r *Resolver) resolveImports(queue *list.List, testDeps, addTest bool) ([]s
 		return []string{}, nil
 	}
 
+	alreadySeen := map[string]bool{}
```
Comment (seh) [inline dependency/resolver.go]: - [ ] Consider using the expected size:
```go
alreadySeen := make(map[string]bool, queue.Len())
```

**Comment by seh** [comment]
I ran _glide cache-clear_, confirmed that all the cached files were gone, and ran _glide install --strip-vendor_ again, and it failed in the same way. At least now it completes its run, rather than getting stuck.
```text
[ERROR]	Error scanning github.com/go-kit/kit/service: open /Users/seh/.glide/cache/src/https-github.com-go-kit-kit/service: no such file or directory
[ERROR]	This error means the referenced package was not found.
[ERROR]	Missing file or directory errors usually occur when multiple packages
[ERROR]	share a common dependency and the first reference encountered by the scanner
[ERROR]	sets the version to one that does not contain a subpackage needed required
[ERROR]	by another package that uses the shared dependency. Try setting a
[ERROR]	version in your glide.yaml that works for all packages that share this
[ERROR]	dependency.
[ERROR]	Error scanning google.golang.org/api/internal: open /Users/seh/.glide/cache/src/https-google.golang.org-api/internal: no such file or directory
[ERROR]	This error means the referenced package was not found.
[ERROR]	Missing file or directory errors usually occur when multiple packages
[ERROR]	share a common dependency and the first reference encountered by the scanner
[ERROR]	sets the version to one that does not contain a subpackage needed required
[ERROR]	by another package that uses the shared dependency. Try setting a
[ERROR]	version in your glide.yaml that works for all packages that share this
[ERROR]	dependency.
```
I had trouble figuring out what's looking for the import path "github.com/go-kit/kit/service". I found [this reference in one of _go-kit_'s integration tests](https://github.com/go-kit/kit/blob/master/sd/consul/integration_test.go#L12). I'm using [the latest release of _go-kit_](https://github.com/go-kit/kit/releases/tag/v0.2.0), but I assume that taking a later version won't fix the problem here, because that _integration_test.go_ file still references this nonexistent package even in the latest version.

**Comment by seh** [comment]
I filed go-kit/kit#366 for the aforementioned problem.

For the second one:
```shell
% find ~/.glide/cache -name '*.go' -type f -exec egrep '\Qgoogle.golang.org/api/internal\E' {} \; -print
	"google.golang.org/api/internal"
/Users/seh/.glide/cache/src/https-camlistore.org/vendor/google.golang.org/api/option/option.go
	"google.golang.org/api/internal"
/Users/seh/.glide/cache/src/https-camlistore.org/vendor/google.golang.org/api/option/option_test.go
	"google.golang.org/api/internal"
/Users/seh/.glide/cache/src/https-camlistore.org/vendor/google.golang.org/api/transport/dial.go
```
The "google.golang.org/api/internal" package does look to be available [here](https://code.googlesource.com/google-api-go-client/+/master/internal/).

**Comment by seh** [comment]
As far as I can tell, this patch does what it intends to do. I'm still plagued by other problems, but apart from [my recommendation for initializing the map's capacity above](https://github.com/Masterminds/glide/pull/622#discussion_r81255399), I recommend merging it.

**Comment by breerly** [comment]
No tests?

**Comment by sdboyer** [comment]
@breerly some of glide's core internals are only partially covered in general. The current resolver has some tests, but not a ton. There's a fair bit of complexity involved in setting up such tests.

...which is not a good argument in general for not having tests. It's a horrible one, in fact. However, it's kinda more acceptable here, because all of this is logic that's also slated to be replaced by gps. And gps has extensive test fixtures covering all manner of weird solver paths.

so, it's less valuable to do relatively complex work here when it's slated to be thrown away in the next version.

**Pull request merged**